### PR TITLE
srml: system: prune block hash mapping

### PR DESCRIPTION
**Pull request by andresilva**
Currently the block number to block hash mapping in the system module grows unbound. This PR adds a module configuration that determines the maximum amount of mappings to keep. On finalize the oldest mapping is pruned. Default is to keep the last 250 blocks. The genesis hash is kept indefinitely.

### CODE REVIEW

**Comment by andresilva** [comment]
Should we special case the genesis block and always keep its hash in the mapping?

File: srml/system/src/lib.rs [hunk 1]
```diff
@@ -317,6 +316,8 @@ decl_storage! {
 		AllExtrinsicsWeight: Option<u32>;
 		/// Map of block numbers to block hashes.
 		pub BlockHash get(block_hash) build(|_| vec![(T::BlockNumber::zero(), hash69())]): map T::BlockNumber => T::Hash;
+		/// Maximum number of block number to block hash mappings to keep (oldest pruned first).
+		BlockHashCount get(block_hash_count) config(): T::BlockNumber = 250.into();
```
Comment (bkchr) [inline srml/system/src/lib.rs]: Sounds like something that should be `parameter_type!`.

File: srml/system/src/lib.rs [hunk 2]
```diff
@@ -580,6 +585,12 @@ impl<T: Trait> Module<T> {
 			digest.push(item);
 		}
 
+		// move block hash pruning window by one block
+		let block_hash_count = <BlockHashCount<T>>::get();
+		if number > block_hash_count {
+			<BlockHash<T>>::remove(number - block_hash_count - One::one());
```
Comment (bkchr) [inline srml/system/src/lib.rs]:
```suggestion
			<BlockHash<T>>::remove(number - block_hash_count);
```

File: srml/system/src/lib.rs [hunk 3]
```diff
@@ -580,6 +585,12 @@ impl<T: Trait> Module<T> {
 			digest.push(item);
 		}
 
+		// move block hash pruning window by one block
+		let block_hash_count = <BlockHashCount<T>>::get();
+		if number > block_hash_count {
```
Comment (bkchr) [inline srml/system/src/lib.rs]:
```suggestion
		if number >+ block_hash_count {
```

Sir, I'm unsure if I'm to tired or not, but I think it should be this way.

Given we have number `101` and `100` as block hash count. This means we already produced `102` blocks.

File: srml/system/src/lib.rs [hunk 4]
```diff
@@ -906,4 +917,45 @@ mod tests {
...
+			// first 5 block hashes are pruned
+			for n in 0..5 {
+				assert_eq!(
+					System::block_hash(n),
+					H256::zero(),
+				);
+			}
+
+			// the remaining 10 are kept
+			for n in 5..15 {
```
Comment (bkchr) [inline srml/system/src/lib.rs]:
```suggestion
			for n in 5..=15 {
```

File: srml/system/src/lib.rs (hunk 1 above)
Comment (andresilva) [reply srml/system/src/lib.rs]: I don't know what `parameter_type!` is, but won't this always be defined in `T::BlockNumber`?

File: srml/system/src/lib.rs (hunk 3 above)
Comment (andresilva) [reply srml/system/src/lib.rs]: If number is `102` then it means that we have just inserted the hash for its parent `101`. And so we will remove `102 - 100 - 1` i.e. block `1`. So our block hash mapping will have the range `[2, 101]` (100 blocks).

File: srml/system/src/lib.rs (hunk 4 above)
Comment (andresilva) [reply srml/system/src/lib.rs]: If we insert until block 15 it means we have only mapped hashes until block 14.

File: srml/system/src/lib.rs (hunk 3 above)
Comment (bkchr) [reply srml/system/src/lib.rs]: I should read the other source code before commenting :D

File: srml/system/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply srml/system/src/lib.rs]: Just look into `runtime` and search for it. You require to always look into the storage for a value that probably never changes. Using `parameter_types!`, the value is hard coded and compiled into the runtime.

**Pull request merged**